pkg/virt-launcher/virtwrap/device/hostdevice/gpu/hostdev.go

```diff
@@ -61,15 +80,138 @@ func CreateHostDevicesFromPools(vmiGPUs []v1.GPU, pciAddressPool, mdevAddressPoo
+	hostDevices := []api.HostDevice{}
+	if !hasGPUsWithDRA(vmi) {
+		return hostDevices, nil
+	}
+	if !isDRADeviceAttributesAvailable(vmi) {
+		return hostDevices, fmt.Errorf("DRA status reconciliation not complete, waiting for %s condition", v1.VirtualMachineInstanceDRAStatusAttributesAvailable)
+	}
+
+	if vmi.Status.DeviceStatus != nil {
+		for _, gpu := range vmi.Status.DeviceStatus.GPUStatuses {
+			if gpu.DeviceResourceClaimStatus != nil && gpu.DeviceResourceClaimStatus.Attributes != nil {
+				// if pciAddress is set, this is a pGPU
+				if gpu.DeviceResourceClaimStatus.Attributes.PCIAddress != nil {
+					continue
+				}
+				// if mdevUUID is set, this is a vGPU
+				if gpu.DeviceResourceClaimStatus.Attributes.MDevUUID != nil {
+					hostDevice := api.HostDevice{
+						Alias: api.NewUserDefinedAlias(AliasPrefix + gpu.Name),
+						Source: api.HostDeviceSource{
+							Address: &api.Address{
+								UUID: *gpu.DeviceResourceClaimStatus.Attributes.MDevUUID,
+							},
+						},
+						Type:  api.HostDeviceMDev,
+						Mode:  "subsystem",
+						Model: "vfio-pci",
+					}
+					gpuSpec := getGPUSpecFromName(vmi, gpu.Name)
+					if gpuSpec != nil && gpuSpec.VirtualGPUOptions != nil {
+						if gpuSpec.VirtualGPUOptions.Display != nil {
+							displayEnabled := gpuSpec.VirtualGPUOptions.Display.Enabled
+							if displayEnabled == nil || *displayEnabled {
+								hostDevice.Display = "on"
+								if gpuSpec.VirtualGPUOptions.Display.RamFB == nil || *gpuSpec.VirtualGPUOptions.Display.RamFB.Enabled {
+									hostDevice.RamFB = "on"
+								}
+							}
+						}
+					}
+					hostDevices = append(hostDevices, hostDevice)
+				}
+			}
+		}
+	}
+	if defaultDisplayOn && !isVgpuDisplaySet(vmi.Spec.Domain.Devices.GPUs) && len(hostDevices) > 0 {
+		hostDevices[0].Display = "on"
+		hostDevices[0].RamFB = "on"
+	}
+	return hostDevices, nil
+}
+
+// hasGPUsWithDRA checks if the VMI has any GPU devices configured with DRA
+func hasGPUsWithDRA(vmi *v1.VirtualMachineInstance) bool {
+	for _, gpu := range vmi.Spec.Domain.Devices.GPUs {
+		if gpu.ClaimRequest != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func isVgpuDisplaySet(gpuSpecs []v1.GPU) bool {
+	for _, gpu := range gpuSpecs {
+		if gpu.VirtualGPUOptions != nil &&
+			gpu.VirtualGPUOptions.Display != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func getGPUSpecFromName(vmi *v1.VirtualMachineInstance, gpu string) *v1.GPU {
+	for _, g := range vmi.Spec.Domain.Devices.GPUs {
+		if g.Name == gpu {
+			return &g
+		}
+	}
+	return nil
+}
+
 func createHostDevicesMetadata(vmiGPUs []v1.GPU) []hostdevice.HostDeviceMetaData {
 	var hostDevicesMetaData []hostdevice.HostDeviceMetaData
 	for _, dev := range vmiGPUs {
-		hostDevicesMetaData = append(hostDevicesMetaData, hostdevice.HostDeviceMetaData{
-			AliasPrefix:       AliasPrefix,
-			Name:              dev.Name,
-			ResourceName:      dev.DeviceName,
-			VirtualGPUOptions: dev.VirtualGPUOptions,
-		})
+		if dev.ClaimRequest == nil {
```

Comment:
WDYT about changing the function name to something like `createDevicePluginHostDevicesMetadata`?

Reply:
done, PTAL